spring-integration-file/src/main/java/org/springframework/integration/file/remote/session/RotatingServerAdvice.java

```diff
@@ -0,0 +1,215 @@
+		 */
+		void afterReceive(boolean messageReceived, MessageSource<?> source);
+
+	}
+
+	/**
+	 * Standard rotation policy; iterates over key/directory pairs; when the end
+	 * is reached, starts again at the beginning. If the fair option is true
+	 * the rotation occurs on every poll, regardless of result. Otherwise rotation
+	 * occurs when the current pair returns no message.
+	 * @since 5.0.7
+	 *
+	 */
+	private static class StandardRotationPolicy implements RotationPolicy {
+
+		private final Log logger = LogFactory.getLog(getClass());
+
+		private final DelegatingSessionFactory<?> factory;
+
+		private final List<KeyDirectory> keyDirectories = new ArrayList<>();
+
+		private final boolean fair;
+
+		private volatile Iterator<KeyDirectory> iterator;
+
+		private volatile KeyDirectory current;
+
+		private volatile boolean initialized;
+
+		StandardRotationPolicy(DelegatingSessionFactory<?> factory, List<KeyDirectory> keyDirectories,
+				boolean fair) {
+			Assert.notNull(factory, "factory cannot be null");
+			Assert.notNull(keyDirectories, "keyDirectories cannot be null");
+			Assert.isTrue(keyDirectories.size() > 0, "At least one KeyDirectory is required");
+			this.factory = factory;
+			this.keyDirectories.addAll(keyDirectories);
+			this.fair = fair;
+			this.iterator = this.keyDirectories.iterator();
+		}
+
+		@Override
+		public void beforeReceive(MessageSource<?> source) {
+			if (this.fair || !this.initialized) {
+				configureSource(source);
+				this.initialized = true;
+			}
+			if (this.logger.isTraceEnabled()) {
+				this.logger.trace("Next poll is for " + this.current);
+			}
+			this.factory.setThreadKey(this.current.getKey());
+		}
+
+		@Override
+		public void afterReceive(boolean messageReceived, MessageSource<?> source) {
+			if (this.logger.isTraceEnabled()) {
+				this.logger.trace("Poll resulted in message: " + messageReceived);
+			}
+			this.factory.clearThreadKey();
+			if (!this.fair && !messageReceived) {
+				configureSource(source);
+			}
+		}
+
+		private void configureSource(MessageSource<?> source) {
+			Assert.isTrue(source instanceof AbstractInboundFileSynchronizingMessageSource
+					|| source instanceof AbstractRemoteFileStreamingMessageSource,
+				"source must be an AbstractInboundFileSynchronizingMessageSource or a "
+				+ "AbstractRemoteFileStreamingMessageSource");
+			if (!this.iterator.hasNext()) {
+				this.iterator = this.keyDirectories.iterator();
+			}
+			this.current = this.iterator.next();
+			if (source instanceof AbstractRemoteFileStreamingMessageSource) {
+				((AbstractRemoteFileStreamingMessageSource<?>) source).setRemoteDirectory(this.current.getDirectory());
+			}
+			else {
+				((AbstractInboundFileSynchronizingMessageSource<?>) source).getSynchronizer()
+					.setRemoteDirectory(this.current.getDirectory());
+			}
+		}
+
+	}
+
+	/**
+	 * A {@link DelegatingSessionFactory} key/directory pair.
+	 * @since 5.0.7
+	 *
+	 */
+	public static class KeyDirectory {
```

Comment:
The `equals()` and `hashCode()` ?

Can't we live with the `Map.Entry` instead BTW ?